File: internal/mirror/mirror.go
// Package mirror turns an opam-repository checkout into a repository whose
// packages point at this server.
//
// The whole idea of opamela lives in this package. An opam repository is not a
// package store: it is a directory of URLs pointing at several thousand
// unrelated servers. That is why putting an HTTP cache in front of one caches
// nothing worth caching, and why the only way to become the single path a build
// takes is to rewrite the directory itself.
package mirror

import (
	"archive/tar"
	"compress/gzip"
	"context"
	"errors"
	"fmt"
	"io"
	"io/fs"
	"net/url"
	"os"
	"path"
	"path/filepath"
	"runtime"
	"sort"
	"strings"
	"sync"
	"sync/atomic"

	"github.com/jeremiegoldberg/opamela/internal/opamfile"
	"github.com/jeremiegoldberg/opamela/internal/opamrepo"
)

// IndexName is the file opam fetches when a repository is served over HTTP.
// Serving the tree alone is not enough, which is a detail worth knowing before
// wondering why a perfectly good directory of opam files does nothing.
const IndexName = "index.tar.gz"

// Stats reports what a build did.
type Stats struct {
	Packages    int // versioned packages seen
	Rewritten   int // now pointing at this server
	Sourceless  int // no url.src: conf-* and virtual packages
	Passthrough int // a source we cannot serve as a file, left untouched
}

func (s Stats) String() string {
	return fmt.Sprintf("%d packages: %d rewritten, %d sourceless, %d passed through",
		s.Packages, s.Rewritten, s.Sourceless, s.Passthrough)
}

// Builder generates a rewritten repository.
type Builder struct {
	// BaseURL is the public root of this server, without a trailing slash,
	// for example "https://opam.internal".
	BaseURL string

	// Workers bounds concurrency. Zero means one per CPU.
	Workers int
}

// Build writes a rewritten copy of the source repositories into dstRoot,
// replacing whatever was there, and generates the index tarball.
//
// Packages are independent, so they are built in parallel. Note what is not
// here: no shared database, no shared handle, nothing to serialise. Each worker
// reads one file and writes another. Concurrency is safe by construction rather
// than by locking, which is the cheapest kind of safe.
func (b *Builder) Build(ctx context.Context, src opamrepo.Sources, dstRoot string) (Stats, error) {
	base := strings.TrimSuffix(b.BaseURL, "/")
	if base == "" {
		return Stats{}, errors.New("mirror: empty BaseURL")
	}

	roots := src.Roots()
	if len(roots) == 0 {
		return Stats{}, errors.New("mirror: no source repository")
	}

	packages, err := opamrepo.List(roots)
	if err != nil {
		return Stats{}, err
	}

	if err := os.RemoveAll(dstRoot); err != nil {
		return Stats{}, err
	}
	if err := os.MkdirAll(dstRoot, 0o755); err != nil {
		return Stats{}, err
	}
	if err := copyRepoFile(src.Base, dstRoot); err != nil {
		return Stats{}, err
	}

	workers := b.Workers
	if workers <= 0 {
		workers = runtime.NumCPU()
	}

	var (
		rewritten, sourceless, passthrough atomic.Int64
		firstErr                           error
		errOnce                            sync.Once
	)

	jobs := make(chan opamrepo.Package)
	var wg sync.WaitGroup
	for i := 0; i < workers; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for pkg := range jobs {
				kind, err := buildPackage(pkg, dstRoot, base)
				if err != nil {
					errOnce.Do(func() { firstErr = err })
					return
				}
				switch kind {
				case rewrittenPackage:
					rewritten.Add(1)
				case sourcelessPackage:
					sourceless.Add(1)
				case passthroughPackage:
					passthrough.Add(1)
				}
			}
		}()
	}

feed:
	for _, pkg := range packages {
		select {
		case jobs <- pkg:
		case <-ctx.Done():
			firstErr = ctx.Err()
			break feed
		}
	}
	close(jobs)
	wg.Wait()

	if firstErr != nil {
		return Stats{}, firstErr
	}

	if err := writeIndex(dstRoot); err != nil {
		return Stats{}, fmt.Errorf("writing %s: %w", IndexName, err)
	}

	return Stats{
		Packages:    len(packages),
		Rewritten:   int(rewritten.Load()),
		Sourceless:  int(sourceless.Load()),
		Passthrough: int(passthrough.Load()),
	}, nil
}

type packageKind int

const (
	rewrittenPackage packageKind = iota
	sourcelessPackage
	passthroughPackage
)

func buildPackage(pkg opamrepo.Package, dstRoot, base string) (packageKind, error) {
	dstDir := filepath.Join(dstRoot, pkg.RelDir())
	if err := os.MkdirAll(dstDir, 0o755); err != nil {
		return 0, err
	}

	// Patches and other side files travel with the package untouched.
	if err := copyExtras(pkg.Dir, dstDir); err != nil {
		return 0, err
	}

	data, err := os.ReadFile(pkg.OpamPath())
	if err != nil {
		return 0, err
	}

	f, _ := opamfile.Parse(data) // Parse never returns an error

	// Rewrite the url source and every mirrorable extra-source in one pass, so
	// that the patches opam fetches at build time also come through the mirror
	// instead of straight from the public internet.
	out, err := opamfile.RewriteSources(data, f, func(s opamfile.Source) string {
		if !Mirrorable(s.Src) {
			return ""
		}
		archive := ArchiveNameFor(s, pkg.Name, pkg.Version)
		return base + DownloadPath(pkg.Name, pkg.Version, archive)
	})
	if err != nil {
		return 0, fmt.Errorf("%s: %w", pkg.OpamPath(), err)
	}

	// Classify the package by its own source, for the build summary. Its
	// extra-sources were rewritten above regardless of this.
	var kind packageKind
	switch {
	case f.URL == nil:
		// conf-* and virtual packages have no source archive.
		kind = sourcelessPackage
	case Mirrorable(f.URL.Src):
		kind = rewrittenPackage
	default:
		// git+https and the like cannot be served as an archive. Leaving the
		// url alone is the honest outcome, and a mirror that pretended
		// otherwise would hand opam a file where it expects a clone.
		kind = passthroughPackage
	}

	// The destination is the versioned package directory. Writing one level
	// up produces packages/<name>/opam, a file opam never reads, and a
	// mirror that silently redirects nothing at all.
	return kind, os.WriteFile(filepath.Join(dstDir, "opam"), out, 0o644)
}

// DownloadPath is the server path a rewritten package points at.
func DownloadPath(pkg, version, archive string) string {
	return "/download/" + pkg + "/" + version + "/" + archive
}

// Mirrorable reports whether a source URL can be served as a plain file.
func Mirrorable(src string) bool {
	u, err := url.Parse(src)
	if err != nil {
		return false
	}
	return (u.Scheme == "http" || u.Scheme == "https") && u.Host != "" && u.Path != ""
}

// ArchiveName is the file name a rewritten package advertises.
//
// It is derived from the upstream URL so that opam's own archive cache keeps
// recognisable names, and it is always a single safe path element: this string
// ends up in a URL, in an opam file and in a filesystem path.
func ArchiveName(src, pkg, version string) string {
	fallback := pkg + "." + version + ".tar.gz"

	u, err := url.Parse(src)
	if err != nil {
		return fallback
	}
	name := path.Base(u.Path)
	if name == "" || name == "." || name == "/" || name == ".." {
		return fallback
	}
	if !safeArchiveName(name) {
		return fallback
	}
	return name
}

// ArchiveNameFor is the file name a rewritten source advertises. For the url
// section it comes from the upstream URL; for an extra-source it is the
// section's label, which opam guarantees to be a filename unique within the
// package. The rare case where an extra-source label collides with the url
// archive name resolves in favour of the url, since the server tries the url
// first; such a collision has not been observed in opam-repository.
func ArchiveNameFor(s opamfile.Source, pkg, version string) string {
	if s.IsExtra() && safeArchiveName(s.Name) {
		return s.Name
	}
	return ArchiveName(s.Src, pkg, version)
}

func safeArchiveName(s string) bool {
	if len(s) > 200 || strings.Contains(s, "..") {
		return false
	}
	for _, r := range s {
		switch {
		case r >= 'a' && r <= 'z', r >= 'A' && r <= 'Z', r >= '0' && r <= '9':
		case r == '.', r == '-', r == '_', r == '+', r == '~':
		default:
			return false
		}
	}
	return true
}

// copyRepoFile carries the repository's own descriptor across, synthesising a
// minimal one if the upstream checkout has none.
func copyRepoFile(srcRoot, dstRoot string) error {
	data, err := os.ReadFile(filepath.Join(srcRoot, "repo"))
	if errors.Is(err, fs.ErrNotExist) {
		data = []byte("opam-version: \"2.0\"\n")
	} else if err != nil {
		return err
	}
	return os.WriteFile(filepath.Join(dstRoot, "repo"), data, 0o644)
}

// copyExtras copies everything in a package directory except its opam file.
func copyExtras(srcDir, dstDir string) error {
	return filepath.WalkDir(srcDir, func(p string, d fs.DirEntry, err error) error {
		if err != nil {
			return err
		}
		rel, err := filepath.Rel(srcDir, p)
		if err != nil {
			return err
		}
		if rel == "." || rel == "opam" {
			return nil
		}
		dst := filepath.Join(dstDir, rel)
		if d.IsDir() {
			return os.MkdirAll(dst, 0o755)
		}
		if !d.Type().IsRegular() {
			return nil
		}
		return copyFile(p, dst)
	})
}

func copyFile(src, dst string) error {
	in, err := os.Open(src)
	if err != nil {
		return err
	}
	defer in.Close()

	if err := os.MkdirAll(filepath.Dir(dst), 0o755); err != nil {
		return err
	}
	out, err := os.OpenFile(dst, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0o644)
	if err != nil {
		return err
	}
	if _, err := io.Copy(out, in); err != nil {
		out.Close()
		return err
	}
	return out.Close()
}

// writeIndex packs the generated tree into index.tar.gz.
//
// The output is deterministic: paths are sorted and timestamps and ownership
// are zeroed. Two builds of the same upstream revision therefore produce byte
// identical tarballs, which makes "did the mirror actually change?" a question
// with an answer.
func writeIndex(root string) error {
	var paths []string
	err := filepath.WalkDir(root, func(p string, d fs.DirEntry, err error) error {
		if err != nil {
			return err
		}
		rel, err := filepath.Rel(root, p)
		if err != nil {
			return err
		}
		if rel == "." || rel == IndexName {
			return nil
		}
		if d.IsDir() || d.Type().IsRegular() {
			paths = append(paths, rel)
		}
		return nil
	})
	if err != nil {
		return err
	}
	sort.Strings(paths)

	tmp := filepath.Join(root, IndexName+".tmp")
	f, err := os.OpenFile(tmp, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0o644)
	if err != nil {
		return err
	}

	gz := gzip.NewWriter(f)
	tw := tar.NewWriter(gz)

	for _, rel := range paths {
		full := filepath.Join(root, rel)
		info, err := os.Lstat(full)
		if err != nil {
			f.Close()
			return err
		}

		hdr := &tar.Header{
			Name:   filepath.ToSlash(rel),
			Mode:   0o644,
			Format: tar.FormatPAX,
		}
		if info.IsDir() {
			hdr.Typeflag, hdr.Mode, hdr.Name = tar.TypeDir, 0o755, hdr.Name+"/"
		} else {
			hdr.Typeflag, hdr.Size = tar.TypeReg, info.Size()
		}
		if err := tw.WriteHeader(hdr); err != nil {
			f.Close()
			return err
		}
		if info.IsDir() {
			continue
		}

		in, err := os.Open(full)
		if err != nil {
			f.Close()
			return err
		}
		if _, err := io.Copy(tw, in); err != nil {
			in.Close()
			f.Close()
			return err
		}
		in.Close()
	}

	if err := tw.Close(); err != nil {
		f.Close()
		return err
	}
	if err := gz.Close(); err != nil {
		f.Close()
		return err
	}
	if err := f.Close(); err != nil {
		return err
	}
	return os.Rename(tmp, filepath.Join(root, IndexName))
}

File: ocaml/lib/mirror.ml
(* Turn opam-repository checkouts into a repository whose packages point at this
   server. Port of the Go package. An opam repository is not a package store: it
   is a directory of URLs pointing at thousands of unrelated hosts. That is why a
   cache in front of one caches nothing worth caching, and why the only way to
   become the single path a build takes is to rewrite the directory itself. *)

let index_name = "index.tar.gz"

type stats = {
  packages : int;
  rewritten : int;
  sourceless : int;
  passthrough : int;
}

let stats_to_string s =
  Printf.sprintf "%d packages: %d rewritten, %d sourceless, %d passed through"
    s.packages s.rewritten s.sourceless s.passthrough

(* --- small helpers ------------------------------------------------------- *)

let rec mkdir_p dir =
  if dir <> "" && dir <> "." && dir <> "/" && not (Sys.file_exists dir) then begin
    mkdir_p (Filename.dirname dir);
    (try Unix.mkdir dir 0o755 with Unix.Unix_error (Unix.EEXIST, _, _) -> ())
  end

let read_file p =
  let ic = open_in_bin p in
  Fun.protect
    ~finally:(fun () -> close_in ic)
    (fun () ->
      let n = in_channel_length ic in
      really_input_string ic n)

let write_file p data =
  mkdir_p (Filename.dirname p);
  let oc = open_out_bin p in
  Fun.protect ~finally:(fun () -> close_out oc) (fun () -> output_string oc data)

let rec rm_rf p =
  match Sys.file_exists p with
  | false -> ()
  | true ->
      if (try Sys.is_directory p with Sys_error _ -> false) then begin
        Array.iter (fun e -> rm_rf (Filename.concat p e)) (Sys.readdir p);
        (try Unix.rmdir p with Unix.Unix_error _ -> ())
      end
      else try Sys.remove p with Sys_error _ -> ()

let run_shell cmd =
  match Unix.system cmd with
  | Unix.WEXITED 0 -> Ok ()
  | Unix.WEXITED n -> Error (Printf.sprintf "command exited %d: %s" n cmd)
  | _ -> Error (Printf.sprintf "command killed: %s" cmd)

(* --- url handling -------------------------------------------------------- *)

let scheme_of src =
  match String.index_opt src ':' with
  | Some i when i + 2 < String.length src && src.[i + 1] = '/' && src.[i + 2] = '/'
    ->
      Some (String.sub src 0 i, String.sub src (i + 3) (String.length src - i - 3))
  | _ -> None

(* whether a source URL can be served as a plain file *)
let mirrorable src =
  match scheme_of src with
  | Some (("http" | "https"), rest) ->
      (match String.index_opt rest '/' with
      | Some i -> i > 0 && i < String.length rest - 1
      | None -> false)
  | _ -> false

let safe_archive_name s =
  if String.length s = 0 || String.length s > 200 then false
  else begin
    let has_dotdot =
      let bad = ref false in
      for i = 0 to String.length s - 2 do
        if s.[i] = '.' && s.[i + 1] = '.' then bad := true
      done;
      !bad
    in
    if has_dotdot then false
    else begin
      let ok = ref true in
      String.iter
        (fun c ->
          match c with
          | 'a' .. 'z' | 'A' .. 'Z' | '0' .. '9' | '.' | '-' | '_' | '+' | '~' -> ()
          | _ -> ok := false)
        s;
      !ok
    end
  end

(* the last path element of a URL, host and query stripped *)
let url_basename src =
  match scheme_of src with
  | None -> ""
  | Some (_, rest) ->
      let rest =
        match String.index_opt rest '?' with
        | Some q -> String.sub rest 0 q
        | None -> rest
      in
      (* rest is host[/path]; keep only from the first '/' *)
      let path =
        match String.index_opt rest '/' with
        | Some i -> String.sub rest i (String.length rest - i)
        | None -> ""
      in
      if path = "" then "" else Filename.basename path

let archive_name src pkg version =
  let fallback = pkg ^ "." ^ version ^ ".tar.gz" in
  let name = url_basename src in
  if name = "" || name = "." || name = "/" || name = ".." then fallback
  else if not (safe_archive_name name) then fallback
  else name

(* The archive name a rewritten source advertises. For the url section it comes
   from the upstream URL; for an extra-source it is its label, which opam
   guarantees unique within a package. *)
let archive_name_for (s : Opamfile.source) pkg version =
  if Opamfile.is_extra s && safe_archive_name s.name then s.name
  else archive_name s.src pkg version

let download_path pkg version archive =
  "/download/" ^ pkg ^ "/" ^ version ^ "/" ^ archive

(* --- building ------------------------------------------------------------ *)

(* copy everything in a package directory except its opam file *)
let rec copy_extras src_dir dst_dir =
  Array.iter
    (fun entry ->
      if entry <> "opam" then begin
        let s = Filename.concat src_dir entry in
        let d = Filename.concat dst_dir entry in
        if (try Sys.is_directory s with Sys_error _ -> false) then begin
          mkdir_p d;
          copy_extras s d
        end
        else if (Unix.lstat s).Unix.st_kind = Unix.S_REG then
          write_file d (read_file s)
      end)
    (Sys.readdir src_dir)

type kind = Rewritten | Sourceless | Passthrough

let build_package (pkg : Opamrepo.package) dst_root base : kind =
  let dst_dir = Filename.concat dst_root (Opamrepo.rel_dir pkg) in
  mkdir_p dst_dir;
  copy_extras pkg.dir dst_dir;

  let data = read_file (Opamrepo.opam_path pkg) in
  let f = Opamfile.parse data in

  (* Rewrite the url source and every mirrorable extra-source in one pass, so
     the patches opam fetches at build time come through the mirror too. *)
  let out =
    Opamfile.rewrite_sources data f (fun (s : Opamfile.source) ->
        if not (mirrorable s.src) then None
        else
          Some (base ^ download_path pkg.name pkg.version (archive_name_for s pkg.name pkg.version)))
  in

  let kind =
    match f.url with
    | None -> Sourceless
    | Some u -> if mirrorable u.src then Rewritten else Passthrough
  in
  write_file (Filename.concat dst_dir "opam") out;
  kind

let copy_repo_file base dst_root =
  let repo = if base = "" then "" else Filename.concat base "repo" in
  let data =
    if repo <> "" && Sys.file_exists repo then read_file repo
    else "opam-version: \"2.0\"\n"
  in
  write_file (Filename.concat dst_root "repo") data

(* Pack the generated tree into index.tar.gz, deterministically: sorted names,
   zeroed timestamps and ownership, gzip without a name/timestamp header. Two
   builds of the same input therefore produce identical tarballs, which makes
   "did the mirror actually change?" a question with an answer. Shelling out to
   tar mirrors how the Go version leans on the standard archive/tar; OCaml's
   stdlib has neither tar nor gzip. *)
let write_index dst_root =
  let cmd =
    Printf.sprintf
      "cd %s && find . -mindepth 1 ! -name %s | LC_ALL=C sort | \
       tar --owner=0 --group=0 --numeric-owner --mtime=@0 --no-recursion \
       -T - -cf - | gzip -n > %s"
      (Filename.quote dst_root) (Filename.quote index_name)
      (Filename.quote index_name)
  in
  run_shell cmd

(* Write a rewritten copy of the source repositories into dst_root, replacing
   whatever was there, and generate the index tarball. base_url is the public
   root this server answers on, without a trailing slash. *)
let build ~base_url (src : Opamrepo.sources) dst_root : (stats, string) result =
  let base_url =
    if String.length base_url > 0 && base_url.[String.length base_url - 1] = '/' then
      String.sub base_url 0 (String.length base_url - 1)
    else base_url
  in
  if base_url = "" then Error "mirror: empty base_url"
  else
    let roots = Opamrepo.roots src in
    if roots = [] then Error "mirror: no source repository"
    else begin
      let packages = Opamrepo.list roots in
      rm_rf dst_root;
      mkdir_p dst_root;
      copy_repo_file src.base dst_root;
      let rewritten = ref 0 and sourceless = ref 0 and passthrough = ref 0 in
      List.iter
        (fun pkg ->
          match build_package pkg dst_root base_url with
          | Rewritten -> incr rewritten
          | Sourceless -> incr sourceless
          | Passthrough -> incr passthrough)
        packages;
    match write_index dst_root with
    | Error e -> Error e
    | Ok () ->
        Ok
          {
            packages = List.length packages;
            rewritten = !rewritten;
            sourceless = !sourceless;
            passthrough = !passthrough;
          }
  end
